### PR TITLE
Eip add runq based overload protection

### PR DESCRIPTION
**Pull request by qzhuyan**

### CODE REVIEW

File: active/0010-runq-based-overload-protection [hunk 1]
```diff
@@ -0,0 +1,120 @@
+# Run queue based overload protection.
+
+## Changelog
+* 2021-03-11: @qzhuyan Initial Draft
+
+## Abstract
+
+Run queue based overload protection for EMQX.
+
+EMQX need some mechanism to cool down the node when the node is overloaded.
```
Comment (k32) [inline active/0010-runq-based-overload-protection:10]: Typo: needs

File: active/0010-runq-based-overload-protection [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+Runq (Erlang VM run queue) is a performance critical metric.
```
Comment (k32) [inline active/0010-runq-based-overload-protection:12]: Nit: performance-critical

File: active/0010-runq-based-overload-protection [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+- Defer processing of non-real-time requests such as Web UI requests.
+
+  consequences:
+
+    Web UI would be slow to respond and operator have to wait.
+
+- Stop spawning more parallel workers.
+
+  consequences:
+    might heart latency.
```
Comment (k32) [inline active/0010-runq-based-overload-protection:70]: Typo: hurt

File: active/0010-runq-based-overload-protection.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+## Motivation
+
+Some user reports EMQX runs at 100% cpu load and etop trace shows the runq of a 2 cores node is around 100
+and lasts more than 5 mins until the user kicks in then restarts the nodes.
+
+Ideally, the runq should not be greater than the number of schedulers. The runq can grow
+for a short period of time to handle peak traffic but should not stay high for longer than 1 mins.
+
+Erlang is soft real-time system, A long run queue can cause time-critical processes not getting scheduled on time that may
```
Comment (k32) [inline active/0010-runq-based-overload-protection.md:25]: Nit: /A long/a long/

File: active/0010-runq-based-overload-protection.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+lead timeout in upper layer's protocol. High level CPU usage may also prevent users from login for O&M operations and leave nodes in zombie state.
```
Comment (k32) [inline active/0010-runq-based-overload-protection.md:26]: Nit: lead to timeouts the upper layer protocol.
"High level" -> just "high"?

File: active/0010-runq-based-overload-protection.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+1. Raise the overload flag by registering a process name such like '__long_runq__'
+
+1. Monitoring systems should be notified such as logging and metrics counter updates.
+
+1. Alert should be triggered, if the flag is not cleared after [X timer] mins.
+
+### When node need cooldown
+
+Each application of EMQX should decide itself what to deal with the overload.
```
Comment (k32) [inline active/0010-runq-based-overload-protection.md:49]: Nit: decide for itself how to deal

**Review by k32 — APPROVED** [review]

File: active/0010-runq-based-overload-protection.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
+# Run queue based overload protection.
+
+## Changelog
+* 2021-03-11: @qzhuyan Initial Draft
+
+## Abstract
+
+Run queue based overload protection for EMQX.
```
[the remark this answers is not in the document]
Comment (qzhuyan) [reply active/0010-runq-based-overload-protection.md:8]: Do you mean run-queue based?

[the remark this answers is not in the document]
Comment (k32) [reply active/0010-runq-based-overload-protection.md:8]: I think hyphen is missing before "based"?

**Pull request merged**